// @grafana/faro-web-tracing/dist/cjs/semconv.js
"use strict";
/**
 Unstable SemConv
 Because the "incubating" entry-point may include breaking changes in minor versions,
 it is recommended that instrumentation libraries not import @opentelemetry/semantic-conventions/incubating in runtime code,
 but instead copy relevant definitions into their own code base. (This is the same recommendation as for other languages.)
 
 See: https://www.npmjs.com/package/@opentelemetry/semantic-conventions#:~:text=%7D)%3B-,Unstable%20SemConv,-Because%20the%20%22incubating
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ATTR_BROWSER_PLATFORM = exports.ATTR_BROWSER_MOBILE = exports.ATTR_BROWSER_LANGUAGE = exports.ATTR_BROWSER_BRANDS = exports.ATTR_TELEMETRY_DISTRO_VERSION = exports.ATTR_TELEMETRY_DISTRO_NAME = exports.ATTR_PROCESS_RUNTIME_VERSION = exports.ATTR_PROCESS_RUNTIME_NAME = exports.ATTR_SERVICE_NAMESPACE = exports.ATTR_DEPLOYMENT_ENVIRONMENT_NAME = exports.ATTR_SESSION_ID = void 0;
exports.ATTR_SESSION_ID = 'session.id';
exports.ATTR_DEPLOYMENT_ENVIRONMENT_NAME = 'deployment.environment.name';
exports.ATTR_SERVICE_NAMESPACE = 'service.namespace';
// https://opentelemetry.io/docs/specs/semconv/resource/process/#javascript-runtimes
exports.ATTR_PROCESS_RUNTIME_NAME = 'process.runtime.name';
exports.ATTR_PROCESS_RUNTIME_VERSION = 'process.runtime.version';
// https://opentelemetry.io/docs/specs/semconv/attributes-registry/telemetry/#telemetry-attributes
exports.ATTR_TELEMETRY_DISTRO_NAME = 'telemetry.distro.name';
exports.ATTR_TELEMETRY_DISTRO_VERSION = 'telemetry.distro.version';
// https://opentelemetry.io/docs/specs/semconv/resource/browser/
exports.ATTR_BROWSER_BRANDS = 'browser.brands';
exports.ATTR_BROWSER_LANGUAGE = 'browser.language';
exports.ATTR_BROWSER_MOBILE = 'browser.mobile';
exports.ATTR_BROWSER_PLATFORM = 'browser.platform';
